lib/src/importer/node_package.dart

```diff
@@ -0,0 +1,398 @@
+  String? _resolvePackageRootValues(
+      String packageRoot, Map<String, dynamic> packageManifest) {
+    if (packageManifest['sass'] case String sassValue) {
+      if (validExtensions.contains(p.extension(sassValue))) {
+        return p.url.join(packageRoot, sassValue);
+      }
+    }
+
+    if (packageManifest['style'] case String styleValue) {
+      if (validExtensions.contains(p.extension(styleValue))) {
+        return p.url.join(packageRoot, styleValue);
+      }
+    }
+
+    var result = resolveImportPath(p.url.join(packageRoot, 'index'));
+    if (result != null) return result;
+    return null;
+  }
+
+  /// Returns a file path specified by a `subpath` in the `exports` section of
+  /// package.json.
+  ///
+  /// `packageName` is used for error reporting.
+  String? _resolvePackageExports(String packageRoot, String? subpath,
+      Map<String, dynamic> packageManifest, String packageName) {
+    if (packageManifest['exports'] == null) return null;
+    var exports = packageManifest['exports'] as Object;
+    var subpathVariants = _exportsToCheck(subpath);
+    if (_nodePackageExportsResolve(
+            packageRoot, subpathVariants, exports, subpath, packageName)
+        case String path) {
+      return path;
+    }
+
+    if (subpath != null && p.extension(subpath).isNotEmpty) return null;
+
+    var subpathIndexVariants = _exportsToCheck(subpath, addIndex: true);
+    if (_nodePackageExportsResolve(
+            packageRoot, subpathIndexVariants, exports, subpath, packageName)
+        case String path) {
+      return path;
+    }
+
+    return null;
+  }
+
+  /// Returns the path to one subpath variants, resolved in the `exports` of a
+  /// package manifest.
+  ///
+  /// Throws an error if multiple `subpathVariants` match, and null if none
+  /// match.
+  ///
+  /// Implementation of `PACKAGE_EXPORTS_RESOLVE` from the [Resolution Algorithm
+  /// Specification](https://nodejs.org/api/esm.html#resolution-algorithm-specification).
+  String? _nodePackageExportsResolve(
+      String packageRoot,
+      List<String?> subpathVariants,
+      Object exports,
+      String? subpath,
+      String packageName) {
+    if (exports is Map<String, dynamic>) {
+      if (exports.keys.any((key) => key.startsWith('.')) &&
+          exports.keys.any((key) => !key.startsWith('.'))) {
+        throw 'Invalid Package Configuration';
+      }
+    }
+    String? processVariant(String? subpath) {
+      if (subpath == null) {
+        return _getMainExport(exports).andThen((mainExport) =>
+            _packageTargetResolve(subpath, mainExport, packageRoot));
+      }
+      if (exports is! Map<String, dynamic> ||
+          exports.keys.every((key) => !key.startsWith('.'))) {
+        return null;
+      }
+      var matchKey = "./$subpath";
+      if (exports.containsKey(matchKey) && !matchKey.contains('*')) {
+        return _packageTargetResolve(
+            matchKey, exports[matchKey] as Object, packageRoot);
```

Comment:
This will currently crash with an unhelpful error if `exports[matchKey]` is null. Same with other `as Object` casts below.

Reply:
Updated in c3b1c8b0ea952119418f3df93b1a068149333c38

Side note, is there a way to add a if case after an && in an existing if statement?